docs/docs/spec.md

```diff
@@ -726,10 +726,10 @@ A type is null-exclusive under every parameterization if it has a
 [nullness-subtype-establishing path] to either of the following:
 
 -   any type whose [nullness operator] is `MINUS_NULL`
--   any augmented class or array type
+-   any augmented class or array type or the augmented null type
```

Comment:
```suggestion
-   any augmented class, array, or null type
```

The `augmented` is supposed to apply to all three, right?
Alternatively, we could add `augmented` to `array type` as well.

Reply:
I have a hard time interpreting this. At first it seems like the "augmented" means that as soon as any class type is reached, including e.g. `Object?`, which seems wrong - how would that establish non-exclusiveness?

However, it is reachable through a `Nullness-subtype-establishing path`... which actually establishes that it is non-null? Would `Non-nullness-subtype-establishing path` be a better way to describe these paths?

Or am I missing something else?

Reply:
Yes, `augmented` applies to all three. I think I was just feeling squeamish about saying "any... null type" as if there are a whole bunch of different null types. But that's not worth making the text more confusing for.

(Also, oops: There _are_ in fact multiple different null types :) And nullness-subtype-establishing path could lead to `null NO_CHANGE`, `null UNSPECIFIED` (in the "some-world" variant of the rule), or `null MINUS_NULL` (if that comes up in practice. That leads into your next question....)

---

Indeed, we don't want a path to `Object?` to establish null-exclusiveness. We avoid that by defining ["Nullness-subtype-establishing direct-supertype edges"](https://jspecify.dev/docs/spec/#nullness-subtype-establishing-direct-supertype-edges) to exist only when the destination node has an appropriate nullness operator. (Similarly, it's important that `Object?` doesn't have a path to itself (an empty path, which we otherwise permit) so that we don't treat _`Object?`_ as null-exclusive. That's handled by the nullness-operator requirement in the definition of ["Nullness-subtype-establishing path."](https://jspecify.dev/docs/spec/#nullness-subtype-establishing-path) It's possible that there's a way to deduplicate the nullness-operator checking between those two sections, even beyond what https://github.com/jspecify/jspecify/pull/602 could achieve.)

---

As for the phrasing: In this section, a "nullness-subtype-establishing path" is used only to establish non-nullness. However, it's also used in the definition of ["Nullness subtyping"](https://jspecify.dev/docs/spec/#nullness-subtyping), where it can establish _subtyping_ between types even when both those types can include null, as in:

```java
@NullMarked
class Foo<T extends @Nullable Object, U extends T> {}
```

In that example, we can determine that `U NO_CHANGE` is a nullness subtype of `T NO_CHANGE` by following the `U`->`T` nullness-subtype-establishing path.